ballista/core/src/execution_plans/shuffle_writer.rs

```diff
@@ -179,40 +138,59 @@ impl ShuffleWriterExec {
         self.stage_id
     }
 
+    pub fn partitions(&self) -> &[usize] {
+        &self.partitions
+    }
+
+    pub fn with_partitions(&self, partitions: Vec<usize>) -> Result<Self> {
+        Self::try_new(
+            self.job_id.clone(),
+            self.stage_id,
+            partitions,
+            self.plan.clone(),
+            self.work_dir.clone(),
+            self.shuffle_output_partitioning.clone(),
+        )
+    }
+
+    pub fn with_work_dir(&self, work_dir: impl Into<String>) -> Result<Self> {
+        Self::try_new(
+            self.job_id.clone(),
+            self.stage_id,
+            self.partitions.to_vec(),
+            self.plan.clone(),
+            work_dir.into(),
+            self.shuffle_output_partitioning.clone(),
+        )
+    }
+
     /// Get the true output partitioning
     pub fn shuffle_output_partitioning(&self) -> Option<&Partitioning> {
         self.shuffle_output_partitioning.as_ref()
     }
 
-    pub fn limit(&self) -> Option<usize> {
-        self.limit
-    }
-
     pub fn execute_shuffle_write(
         &self,
-        input_partition: usize,
         context: Arc<TaskContext>,
     ) -> impl Future<Output = Result<Vec<ShuffleWritePartition>>> {
         let mut path = PathBuf::from(&self.work_dir);
         path.push(&self.job_id);
         path.push(&format!("{}", self.stage_id));
 
-        let write_metrics = ShuffleWriteMetrics::new(input_partition, &self.metrics);
+        let write_metrics = ShuffleWriteMetrics::new(&self.metrics);
         let output_partitioning = self.shuffle_output_partitioning.clone();
         let plan = self.plan.clone();
 
-        let limit_and_accumulator = self
-            .limit
-            .map(|l| (l, get_limit_accumulator(&self.job_id, self.stage_id)));
+        let partitions = self.partitions.clone();
 
         async move {
             let now = Instant::now();
-            let mut stream = plan.execute(input_partition, context)?;
+            let mut stream = plan.execute(0, context)?;
 
             match output_partitioning {
                 None => {
                     let timer = write_metrics.write_time.timer();
-                    path.push(&format!("{input_partition}"));
+                    path.push(&partitions.iter().map(|p| p.to_string()).join("_"));
```

Comment:
Note: here the partitions are used for shuffle files (`1_2_3_4`) which is probably helpful.